Web socket notifications for state changes.

ebaauw opened the pull request:
Multiple fixes in the handling of web socket notifications for `state` changes:
- Report `orientation` as an array in web socket notification (as is done in the GET response), see #748;
- Combine multiple `state` changes for a light or group into a single web socket notification (as is done for sensors), see #973 issue 5;
- Report `xy` as an array in the web socket notification (as is done in the GET response), see #973 issue 5;
- Use CIE 1931 values (between 0 and 0.9961) for `xy` in the web socket notification (as is done in the GET response), see #973 issue 4;
- Round `xy` values to four decimals (as is done by the Hue bridge), see #973 issue 1.

@manup, I propagated `lastStatePush` from `Sensor` to `RestNodeBase` and added `RestNodeBase` as super type of `Group`.

Kane610 commented:
> Report xy as an array in the web socket notification (as is done in the GET response), see #973 issue 5;
This looks like a breaking change

ebaauw commented:
> This looks like a breaking change

I suppose so: if you rely solely on web socket notifications, you no longer get to see `xy` values.  I think any app should be able to handle (i.e. ignore) new attributes.

I could remove one `else` from the code and (for the time being?) include in the web socket notification both the new `xy` attribute (with the normalised floating point values) as well as the old `x` and `y` attributes (with the raw integer Zigbee values).  Would that help?

ebaauw commented:
And maybe update the API version?  Personally, I don't use this anymore - I sort of lost track what it's supposed to mean when it was bumped from 1.0.9 to 1.16.0 in https://github.com/dresden-elektronik/deconz-rest-plugin/commit/f85529781bb09894ec07d4b021deeb4415d84979.

Kane610 commented:
Doing it gradually would be really great. This is the first breaking change, but maybe there should be an explicit strategy for that? Maybe keep a behaviour for minimum of 2 releases or something.

Kane610 commented:
Just to be clear; I do like this effort. It is just hard to avoid bad experiences among users, if it is just removed without giving some time to upgrade (in this case, two systems)

ebaauw commented:
Not to worry, @Kane610.  New commit includes both `xy` and `x` and `y` in the message.

manup commented:
Sorry for the delay, thanks a lot for the good work. Here are a few comments for the PR `lastStatePush`.

>@manup, I propagated lastStatePush from Sensor to RestNodeBase and added RestNodeBase as super type of Group.

I'm afraid RestNodeBase is not a good fit here, since this base class originally was intended for nodes, aka devices, only. I would love to get rid of RestNodeBase completely in mid future in favour of Resources and ResourceItems.

It would be great if you can add the `lastStatePush` timestamp in form of a ResourceItem. Since everything including groups are already Resources, this is the most general and future proof method.

ebaauw commented:
> It would be great if you can add the `lastStatePush` timestamp in form of a ResourceItem.

I’m not sure that would be a good idea.  We don’t want to expose `lastStatePush` through the REST API.  So we’d need additional code in rest_*.cpp to ignore this ResourceItem, on GET as well as for the web socket notifications (basically everywhere where we loop over a resource’s ResourceItems).  If you want to go this route, we’d better introduce a `hidden` or `notExposed` property for the ResourceItem.

ebaauw commented:
@manup, I moved `lastStatePush` from `RestNodeBase` to `Resource` and removed `RestNodeBase` as superclass for `Group`.  It's a push of the resource's state - I don't know why I didn't put it there in the first place.

ebaauw commented:
Added a bug fix for value of `state.current` for the Heiman smartplug.
